### ISSUE TITLE
Puma can hang on IO.select

### ISSUE THREAD

**Issue by sofianegargouri** [puma/puma#1628]
### Steps to reproduce

1) Create Rails app

2) Deploy it following this tutorial: https://www.digitalocean.com/community/tutorials/deploying-a-rails-app-on-ubuntu-14-04-with-capistrano-nginx-and-puma

### Expected behavior

Work nice and smoothly

### Actual behavior

The server works well for a few hours/days, then it suddenly "stops".
`htop` shows that Puma is still working, but Nginx gives this error:

```
2018/08/04 09:55:00 [error] 889#889: *1241 connect() to unix:///home/app/shared/tmp/sockets/app-puma.sock failed (111: Connection refused) while connecting to upstream, client: 197.14.151.135, server: server.com, request: "GET /my_route HTTP/1.1", upstream: "http://unix:///home/app/shared/tmp/sockets/app-puma.sock:/500.html", host: "server.com"
```

When doing `ls /home/app/shared/tmp/sockets/`, `app-puma.sock` is present

And there is only this in `puma.error.log`:

```
=== puma startup: 2018-07-19 10:13:17 +0000 ===
I, [2018-07-19T10:13:19.790942 #32107]  INFO -- sentry: ** [Raven] Raven 2.7.3 ready to catch errors
* Listening on unix:///home/app/shared/tmp/sockets/app-puma.sock
* Restarting...
Refreshing Gemfile
Puma starting in single mode...
* Version 3.11.3 (ruby 2.4.4-p296), codename: Love Song
* Min threads: 4, max threads: 16
* Environment: production
* Daemonizing...
```

To fix it, I can't even use `cap production deploy:restart`. I have to login on the server, `kill -9 <puma_pid>` and then restart the app from Capistrano.

Honestly I have no idea where this is coming from. I tried it on different servers with the same problem.

I first thought it was my server "stats" (2 GB memory | 1 vCPU | 50 GB SSD) that were low, but it makes no sense since the app is running locally on my Rpi.

### System configuration

**Ruby version**: ruby 2.4.4p296 (2018-03-28 revision 63013) [x86_64-linux]
**Rails version**: 5.1.6
**Puma version**: 3.12.0

**Comment by ce07c3** [comment]
Same here.

**Comment by alexdowad** [comment]
When a process (not just Puma, but any process) freezes and can't be stopped without `kill -9`, what I personally do is:

- `sudo gdb --pid=<PID of defunct process>`
- `thread apply all bt`

...That will show you what each thread is doing. If you can't make sense of the output, copy the whole thing (each page) and share it here. Usually you will find that some thread is stuck trying to do an I/O operation which never completes, and the name of the C function which it is in will give you a clue what kind of I/O it is (perhaps trying to talk to another process).

You need to have GDB installed on your server. Also, your Ruby interpreter must be built with debugging symbols (but I have always found that to be the case).

**Comment by monteirobrena** [comment]
I've been the same error, the application stop and I need to restart Puma always.

**Comment by 1c7** [comment]
if you guys don't have time deal with these.
use Docker and a simple healthcheck. so Docker would restart it for you.

**Comment by nateberkopec** [comment]
Sounds valid but we've got literally nothing to go on here.

**Comment by monteirobrena** [comment]
I don't find a way to fix it. I migrate my application from Azure to AWS and change the Puma to Passenger and now everything works fine. No job stop to work anymore.

**Comment by zentetsukenz** [comment]
I seems to be able to dig down to the cause of the hang.

TL;DR The hang is from `Kernel#select` in `NIO` gems. The `TCPSocket` is not ready for some reason. Hence, timeout when `Kernel#select`.

**Context:**

In my machine (MacOSX 10.14.6), Puma hangs after receiving a certain amount of requests. The number of requests is random. Usually less than 200 requests, around ~150 - 170. My only fix is to restart the host then Puma become healthy. I run Puma in a single mode (1 worker), with min/max 5 threads.

At first I thought it was something inside Puma but after prying everywhere in Puma server/event/etc. 😂 In my case, I found that Puma hangs at [selector.select](https://github.com/puma/puma/blob/v4.1.0/lib/puma/reactor.rb#L136) in `Reactor#run_internal` which is using `NIO::Selector` internally.

At this point it seems like this `stop the world bug` has nothing to do with Puma but I think I should post here first in case someone here know how is it possible that `IO.select` with `TCPServer` socket can hang.

Continue.

I've been messing around with `NIO` until I find a way to change from using `libev` backend to native Ruby. Finally, I found the line that cause the hang [Kernel#select](https://github.com/socketry/nio4r/blob/v2.5.1/lib/nio/selector.rb#L92). It stops since all readers IO are not ready for some reasons.

My current knowledge ends here. If anyone can point me to the right direction to debug this I would really appreciated.

PS. I think the cause be from my machine setting not the Puma or even `Kernel#select` but I would like to know why.

**Comment by nateberkopec** [comment]
@ioquatix does that explanation jive with you?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ioquatix** [comment]
At the very least there are two unrelated symptoms here, because the original issue was reported a year ago before puma use nio. So I can’t comment much more then that without some more details.

Regarding select hang hypothesis can you use strace on Linux to capture log of system calls?

**Comment by zentetsukenz** [comment]
@ioquatix I'm on the MacOSX so I use `dtruss` instead. And to be honest, I'm not even sure the correct way to use it.

Anyway, I did inspection using `sudo dtruss -p <pid> -f` and try to compare the logs between working and not working case. The whole log is too overwhelm so I'll only put the different here.

**When `Kernel.select` hangs.**

```
36651/0x304c2:  write_nocancel(0x1, "\b\b\b\bKernel.select(readers, writers, [], timeout)\0", 0x30)                 = 48 0
36651/0x304c2:  sigprocmask(0x1, 0x0, 0x70000237CF40)            = 0x0 0
36651/0x304c2:  sigaltstack(0x0, 0x70000237CF30, 0x0)            = 0 0
36651/0x304c2:  read(0x0, "\r\0", 0x1)           = 1 0
36651/0x304c2:  write_nocancel(0x1, "\n\0", 0x1)                 = 1 0
36651/0x304c2:  ioctl(0x0, 0x80487415, 0x10895CEE8)              = 0 0
36651/0x304c2:  sigaction(0x1C, 0x70000237D168, 0x70000237D190)          = 0 0
36651/0x304c2:  ioctl(0x24, 0x4004667A, 0x70000237DAC4)          = 0 0
36651/0x304c2:  lstat64("/Users\0", 0x70000237D6B8, 0x0)                 = 0 0

...
There are a bunch of open syscall here.
...

36651/0x304c2:  fstat64(0x25, 0x70000237D8E8, 0x0)               = 0 0
36651/0x304c2:  close(0x25)              = 0 0
36651/0x304c2:  ioctl(0x1, 0x4004667A, 0x70000237DAC4)           = 0 0
36651/0x304c2:  ioctl(0x1, 0x40087468, 0x70000237DB20)           = 0 0
```

I'm not sure that `write_nocancel` is when I enter `Kernel.select` or not but I assume it's that when I enter the command. Between `lstat64` to the next `fstat64` there are a bunch of `open` syscalls to project files and rbenv gem file. First, it return code 0 0 and then returns -1 Err#2 . Let me know if you want more information on this. I'll have to obscure the path in my project btw.

**But when it's working**

```
8511/0x803d:  write(0x17, "Kernel.select(readers, writers, [], timeout)\0", 0x2C)          = 44 0
8511/0x803d:  sigprocmask(0x1, 0x0, 0x7000053F4000)             = 0x0 0
8511/0x803d:  sigaltstack(0x0, 0x7000053F3FF0, 0x0)             = 0 0
8511/0x803d:  write(0x17, "\n\0", 0x1)          = 1 0
8511/0x803d:  sigprocmask(0x1, 0x0, 0x7000053F3EF0)             = 0x0 0
8511/0x803d:  sigaltstack(0x0, 0x7000053F3EE0, 0x0)             = 0 0
8511/0x803d:  select(0x15, 0x7FB0BDF463E0, 0x7FB0BDF46460, 0x7FB0BDF6C3B0, 0x7000053F3F58)                 = 1 0
8511/0x803d:  lstat64("/Users\0", 0x7000053F37B8, 0x0)          = 0 0

...
There are a bunch of open syscall here.
...

8511/0x803d:  fstat64(0x18, 0x7000053F39E8, 0x0)                = 0 0
8511/0x803d:  close(0x18)               = 0 0
8511/0x803d:  ioctl(0x1, 0x4004667A, 0x7000053F3BC4)            = 0 0
8511/0x803d:  ioctl(0x1, 0x40087468, 0x7000053F3C20)            = 0 0
```

Notice that there is the syscall to `select` in the log when it's working.

Not sure this could help much. Let me know if you want more information.

Note. I'm looking into `Kernel#select` now, not sure why `select` is not called.

**Comment by ioquatix** [comment]
Where does the Ruby code call `Kernel.select`, is that in puma?

**Comment by MSP-Greg** [comment]
> is that in puma?

From the link above:

[`lib/nio/selector.rb:92`](https://github.com/socketry/nio4r/blob/v2.5.1/lib/nio/selector.rb#L92)

**Comment by ioquatix** [comment]
Ah, I see.

So, Puma is hanging, both with NIO4r and possibly the original implementation? So, maybe Puma is calling read when there is no possibility of additional data and this hangs?

**Comment by ioquatix** [comment]
I guess we need to see a backtrace of where it is, and what it's expecting to read.

**Comment by zentetsukenz** [comment]
This is what I understand from the Puma implementation.

1. Puma start TCPServer instance when booted. If I remember correctly, Puma keeps the TCPServer in Puma Client object.

2. In Puma Server class, there is a loop constantly read from sockets. [Puma#handle_servers](https://github.com/puma/puma/blob/v4.1.0/lib/puma/server.rb#L378).

3. While idle, the [IO#select](https://github.com/puma/puma/blob/v4.1.0/lib/puma/server.rb#L380) returns only two IO objects which I guess they're file descriptors pipe.

4. When request comes, the third element is present, `TCPServer`. At this point, the IO can select this TCPServer socket without a problem even if it has zero size. And to my understanding, those zero size TCPServer is there just to represent a handshake packet. i.e., SYN, ACK.

I might be wrong with the flow above since debugging this reading loop is a bit hard to me.

5. I don't remember how Client called `try_to_finish` method but there is something to do with re `Puma::Reactor#run_internal` method. Which is using `NIO#select`. Please note that Mac host uses `libev` by default, you have to set env `NIO4R_PURE` to be able to use native ruby which is easier to debug for me 😂.

And yeah, that's the [Hang line](https://github.com/puma/puma/blob/v4.1.0/lib/puma/reactor.rb#L136) in Puma source.

To emphasize, what it tries to read is the `TCPServer` file descriptor. And from my observation, when it hangs, the size of the file is always zero. However, not all zero `TCPServer` file descriptors can't be read.

Please note that "Hang" in my term means that `IO.select` tries to read a file descriptor until it reaches timeout. That's why Puma is not hang since there is always a timeout when trying to read from socket.

**Comment by zentetsukenz** [comment]
Adding to the comment above, `libev` is also hang, that's why I started this whole debugging.

**Comment by tartakynov** [comment]
It could be that you're hitting max open files limit if your service opens too many connections to upstream services or keeps to many files open. In that case puma will not raise any error but silently reject/timeout connections.

To test that you can run your service with a single puma worker and check open files with `lsof` and try increasing `ulimit`

**Comment by ioquatix** [comment]
You should never use `IO.select` in production code. It has too many limits and performance issues.

**Comment by nateberkopec** [comment]
Closing, not enough to go on here. If someone can produce a case that reliably hangs Puma at a certain spot, we can reopen.
